keps/sig-scheduling/5004-dra-extended-resource/README.md

```diff
@@ -0,0 +1,995 @@
+    requests for foo1.domain/bar and foo2.domain/bar, the allocation of devices
+    for each are recorded in DeviceResourceRequestAllocationResult, and
+    one claim object with two `Results` is created for the pod.
+
+The special resource claim object lifecycle is managed by the scheduler and
+garbage collector.
+
+  * It is *created* in a namespace when there is a pod with extended resource
+  request, and the extended resource is advertised by `ResourceSlice` and
+  scheduler has fit the pod to a node with the `ResourceSlice`.
+  * It is *created* by the scheduler dynamic resource plugin during
+    preBind phase. The in-memory one in the assumed cache is created earlier
+    during Reserve phase.
+  * It is *deleted*
+    * either together with the owning pod's deletion.
+    * or by the scheduler dynamic resource plugin during unReserve phase.
+  * It is *read* by the kubelet DRA device driver to prepare the devices listed
+    therein when preparing to run the pod.
+
+```go
+type DeviceRequest struct {
+	// Name can be used to reference this request in a pod.spec.containers[].resources.claims
+	// entry and in a constraint of the claim.
+	//
+	// Must be a DNS label.
+	//
+	// +required
+	Name string `json:"name" protobuf:"bytes,1,name=name"`
+}
+```
+
+As shown above, device request must have a name, and it must be a DNS label.
+kube-scheduler creates one `DeviceRequest` per extended resource backed by DRA
+per container in the pod. The name of the `DeviceRequest` is determined by the
+container name, and the extended resource backed by DRA name. It is in the
+form of "c%d-e%d", where the first %d is the index of the container in the pod.
+The second %d is the index of the extended resource inside the container
+resource requests. For example, if the first container in the pod has an
+extended resource backed by DRA which is the 3rd such request in the container,
+then the name of the `DeviceRequest` is "c0-e2".
+
+### Pod API
+
+A new field `extendedResourceClaimStatus` is added to Pod's status to track
+the special resouceclaim object created for the extended resource requests
+in the pod. This is needed for kublet to pass the devices allocated by driver
+to the containers in the pod.
+
+```go
+// PodExtendedResourceClaimStatus is stored in the PodStatus for each extended
+// resource requests backed by DRA. It stores the generated name for 
+// the corresponding special ResourceClaim created by scheduler.
+type PodExtendedResourceClaimStatus struct {
+        // Names identifies the mapping of <container, extended resource backed by DRA> to  device request.
+        // +patchMergeKey=requestName
+        // +patchStrategy=merge,retainKeys
+        // +listType=map
+        // +listMapKey=requestName
+        // +featureGate=DynamicResourceAllocation
+        Names []ContainerExtendedResourceRequest `json:"names" patchStrategy:"merge,retainKeys" patchMergeKey:"requestName" protobuf:"bytes,1,rep,name=names"`
```

Comment:
I am not sure what this field is needed for.

If we use the `container-%d-request-%d` naming specified above, then the kubelet can parse that to determine the container.

If the purpose of this field is to avoid such string parsing, then it needs to be called out above that the request names are arbitrary and that the `container-%d-request-%d` scheme is not required. This is probably cleaner.

If we keep this struct as-is for that reason, then make this a `+listType=atomic`. It's simpler and more efficient (no need to track individual owners in managed fields when there is always only exactly one owner).

Reply:
This wasn't clear above. Let's clarify as suggested in https://github.com/kubernetes/enhancements/pull/5136/files#r2073542772